README
NAME
    File::pushd - change directory temporarily for a limited scope

VERSION
    This documentation describes version 1.01.

SYNOPSIS
      use File::pushd;
     
  chdir $ENV{HOME};
     
  # change directory again for a limited scope
      {
          my $dir = pushd( '/tmp' );
          # working directory changed to /tmp
      }
      # working directory has reverted to $ENV{HOME}
     
  # tempd() is equivalent to pushd( File::Temp::tempdir )
      {
          my $dir = tempd();
      }
     
  # object stringifies naturally as an absolute path
      {
         my $dir = pushd( '/tmp' );
         my $filename = File::Spec->catfile( $dir, "somefile.txt" );
         # gives /tmp/somefile.txt
      }

DESCRIPTION
    File::pushd does a temporary "chdir" that is easily and automatically
    reverted, similar to "pushd" in some Unix command shells. It works by
    creating an object that caches the original working directory. When the
    object is destroyed, the destructor calls "chdir" to revert to the
    original working directory. By storing the object in a lexical variable
    with a limited scope, this happens automatically at the end of the
    scope.

    This is very handy when working with temporary directories for tasks
    like testing; a function is provided to streamline getting a temporary
    directory from File::Temp.

    For convenience, the object stringifies as the canonical form of the
    absolute pathname of the directory entered.

USAGE
      use File::pushd;

    Using File::pushd automatically imports the "pushd" and "tempd"
    functions.

  pushd
      {
          my $dir = pushd( $target_directory );
      }

    Caches the current working directory, calls "chdir" to change to the
    target directory, and returns a File::pushd object. When the object is
    destroyed, the working directory reverts to the original directory.

    The provided target directory can be a relative or absolute path. If
    called with no arguments, it uses the current directory as its target
    and returns to the current directory when the object is destroyed.

    If the target directory does not exist or if the directory change fails
    for some reason, "pushd" will die with an error message.

  tempd
      {
          my $dir = tempd();
      }

    This function is like "pushd" but automatically creates and calls
    "chdir" to a temporary directory created by File::Temp. Unlike normal
    File::Temp cleanup which happens at the end of the program, this
    temporary directory is removed when the object is destroyed. (But also
    see "preserve".) A warning will be issued if the directory cannot be
    removed.

    As with "pushd", "tempd" will die if "chdir" fails.

  preserve
      {
          my $dir = tempd();
          $dir->preserve;      # mark to preserve at end of scope
          $dir->preserve(0);   # mark to delete at end of scope
      }

    Controls whether a temporary directory will be cleaned up when the
    object is destroyed. With no arguments, "preserve" sets the directory to
    be preserved. With an argument, the directory will be preserved if the
    argument is true, or marked for cleanup if the argument is false. Only
    "tempd" objects may be marked for cleanup. (Target directories to
    "pushd" are always preserved.) "preserve" returns true if the directory
    will be preserved, and false otherwise.

SEE ALSO
    *   File::chdir

BUGS
    Please report any bugs or feature using the CPAN Request Tracker. Bugs
    can be submitted through the web interface at
    <http://rt.cpan.org/Dist/Display.html?Queue=File-pushd>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    David A. Golden (DAGOLDEN)

COPYRIGHT AND LICENSE
    Copyright (c) 2005, 2006, 2007 by David A. Golden

    Licensed under the Apache License, Version 2.0 (the "License"); you may
    not use this file except in compliance with the License. You may obtain
    a copy of the License at <http://www.apache.org/licenses/LICENSE-2.0>

    Files produced as output though the use of this software, including
    generated copies of boilerplate templates provided with this software,
    shall not be considered Derivative Works, but shall be considered the
    original work of the Licensor.

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



Changes
Revision history for Perl module File::pushd

{{$NEXT}}

1.016     2018-05-20 13:31:17-04:00 America/New_York

    - No changes from 1.015-TRIAL

1.015     2018-05-13 19:26:26-04:00 America/New_York (TRIAL RELEASE)

    [FIXED]

    - Directories created with tempd will only be cleaned up by the
      PID that created them, not by forked children.

1.014     2016-10-09 23:26:19-04:00 America/New_York

    - No changes from 1.013-TRIAL.

1.013     2016-10-05 12:57:44-04:00 America/New_York (TRIAL RELEASE)

    [TESTING]

    - Fixed test failures on some 5.8 perls.

1.011     2016-10-05 12:35:27-04:00 America/New_York (TRIAL RELEASE)

    [CHANGED]

    - pushd/tempd warn if called in void context (GitHub #9 Olivier Mengué)

1.009     2014-07-03 13:17:48-04:00 America/New_York

    [FIXED]

    - s/0.95/0.96/ in t/File_pushd.t

1.008     2014-07-03 13:03:30-04:00 America/New_York

    [FIXED]

    - Require a newer version of Test::More

1.007     2014-05-30 23:59:02-04:00 America/New_York

    [FIXED]

    - Fixed a testing edge case on Windows

1.006     2014-03-27 15:29:36-04:00 America/New_York

    [DOCUMENTED]

    - Noted that multiple pushd objects in a lexical scope are not
      destroyed in a predictable order

    [CHANGED]

    - Modernized Dist::Zilla configuration and repo support files

1.005     2013-03-22 09:58:20 America/New_York

    [BUG FIXES]

    - Fix failing test for changing to non-existent directory on some
      older Perls for some platforms

1.004     2013-03-05 17:59:08 America/New_York

    [BUG FIXES]

    - Object destruction won't clobber $@

    [SUPPORT]

    - Changed issue tracker to Github

    - Dropped support for Perl 5.005

1.003     2012-11-27 17:49:03 America/New_York

    [BUG FIXES]

    - Fix test failures on Win32 when using perl.exe that must link to gcc
      DLL [rt.cpan.org #81476]

1.002     2011-12-07 11:40:26 EST5EDT

    [BUG FIXES]

    - Fix failures on cygwin [rt.cpan.org #73072 Reini Urban]

1.001     2011-09-14 22:49:56 America/New_York

    - updated docs to clarify that pushd() and tempd() croak on errors
    - if tempd() croaks, error appears at tempd()'s caller, not from within
      tempd() itself
    - warnings enabled if perl version >= 5.006
    - adds 'untaint_pattern' option with a passable default
    
1.00 Tue Sep 25 08:19:30 EDT 2007
    - changed to the Apache License, version 2.0; (it's clearer, relicensable,
      and is explicit about contributions)
    - moved pod/coverage tests to xt/ 
    - preserve() will preserve a temporary directory beyond the program end
      (by design rather than due to bug RT#21953 in File::Temp); added 
      tests to confirm this behavior after File::Temp is fixed

0.99 Sat Oct 21 08:56:07 EDT 2006
    - bumped version for pre-release before 1.00
    - updated Pod and converted to wikidoc style
    - added example directory

0.32 Sat Feb  4 23:41:33 EST 2006
    - eliminated syntax that prevented backwards compatibility to 5.004
    - adjusted Build.PL to lesser dependencies
    - added optional pod/pod-coverage test (skipped by default)
    - switching to development-odd/release-even version numbering
    
0.30 Sat Dec  3 10:24:16 EST 2005
    - eliminated subclassing of Path::Class to eliminate this non-core
      dependency and improve cross platform compatibility
    - improved portability of tests

0.24 Tue Nov  8 09:40:02 EST 2005
    - added Exporter to @ISA for backwards compatibility [Anton Berezin]
    
0.23 Mon Nov  7 22:14:51 EST 2005
    - derived directories revert to Path::Class::Dir objects instead of
      File::pushd objects

0.22 Mon Nov  7 18:26:50 EST 2005
    - fixed bug in Win32 tests for changing to root directory
      (work around a Path::Class/File::Spec bug)
      
0.21 Mon Nov  7 15:08:17 EST 2005
    - objects created are now subclasses of Path::Class::Dir
    - added ability to preserve temporary directories after creation
    - fixed overloading bug (added fallback)

0.20 Fri Sep  2 06:47:07 EDT 2005
    - added tempd function

0.10 Thu Sep  1 23:53:46 EDT 2005
	- Initial release


Build.PL
use strict;
use Module::Build;
use File::Spec;

my $class = Module::Build->subclass(
    class => "Module::Build::WikiDoc",
    code => <<'SUBCLASS', );

    sub ACTION_wikidoc {
        my $self = shift;
        eval "use Pod::WikiDoc";
        if ( $@ eq '' ) {
            my $parser = Pod::WikiDoc->new({ 
                comment_blocks => 1,
                keywords => { VERSION => $self->dist_version },
            });
            for my $src ( keys %{ $self->find_pm_files() } ) {
                (my $tgt = $src) =~ s{\.pm$}{.pod};
                $parser->filter( {
                    input   => $src,
                    output  => $tgt,
                });
                print "Creating $tgt\n";
                $tgt =~ s{\\}{/}g;
                $self->_add_to_manifest( 'MANIFEST', $tgt );
            }
        }
        else {
            warn "Pod::WikiDoc not available. Skipping wikidoc.\n";
        }
    }

    sub ACTION_test {
        my $self = shift;
        my $missing_pod;
        for my $src ( keys %{ $self->find_pm_files() } ) {
            (my $tgt = $src) =~ s{\.pm$}{.pod};
            $missing_pod = 1 if ! -e $tgt;
        }
        if ( $missing_pod ) {
            $self->depends_on('wikidoc');
            $self->depends_on('build');
        }
        $self->SUPER::ACTION_test;
    }

    sub ACTION_testpod {
        my $self = shift;
        $self->depends_on('wikidoc');
        $self->SUPER::ACTION_testpod;
    }

    sub ACTION_distdir {
        my $self = shift;
        $self->depends_on('wikidoc');
        $self->SUPER::ACTION_distdir;
    }

SUBCLASS

$class->new( 
    module_name         => 'File::pushd',
    dist_author         => 'David A Golden <dagolden@cpan.org>',
    license             => 'apache',
    create_readme       => 1,
    create_makefile_pl  => 'traditional',
    requires            => {
        'perl'          => 5.004,
        'Cwd'           => 0,
        'File::Path'    => 0,
        'File::Spec'    => 0,
        'File::Temp'    => 0,
        'Test::More'    => 0.45, # thread safe
    },
)->create_build_script;



lib/File/pushd.pm
use strict;
use warnings;

package File::pushd;
# ABSTRACT: change directory temporarily for a limited scope

our $VERSION = '1.017';

our @EXPORT = qw( pushd tempd );
our @ISA    = qw( Exporter );

use Exporter;
use Carp;
use Cwd qw( getcwd abs_path );
use File::Path qw( rmtree );
use File::Temp qw();
use File::Spec;

use overload
  q{""}    => sub { File::Spec->canonpath( $_[0]->{_pushd} ) },
  fallback => 1;

#--------------------------------------------------------------------------#
# pushd()
#--------------------------------------------------------------------------#

sub pushd {
    # Called in void context?
    unless (defined wantarray) {
        warnings::warnif(void => 'Useless use of File::pushd::pushd in void context');
        return
    }

    my ( $target_dir, $options ) = @_;
    $options->{untaint_pattern} ||= qr{^([-+@\w./]+)$};

    $target_dir = "." unless defined $target_dir;
    croak "Can't locate directory $target_dir" unless -d $target_dir;

    my $tainted_orig = getcwd;
    my $orig;
    if ( $tainted_orig =~ $options->{untaint_pattern} ) {
        $orig = $1;
    }
    else {
        $orig = $tainted_orig;
    }

    my $tainted_dest;
    eval { $tainted_dest = $target_dir ? abs_path($target_dir) : $orig };
    croak "Can't locate absolute path for $target_dir: $@" if $@;

    my $dest;
    if ( $tainted_dest =~ $options->{untaint_pattern} ) {
        $dest = $1;
    }
    else {
        $dest = $tainted_dest;
    }

    if ( $dest ne $orig ) {
        chdir $dest or croak "Can't chdir to $dest\: $!";
    }

    my $self = bless {
        _pushd    => $dest,
        _original => $orig
      },
      __PACKAGE__;

    return $self;
}

#--------------------------------------------------------------------------#
# tempd()
#--------------------------------------------------------------------------#

sub tempd {
    # Called in void context?
    unless (defined wantarray) {
        warnings::warnif(void => 'Useless use of File::pushd::tempd in void context');
        return
    }

    my ($options) = @_;
    my $dir;
    eval { $dir = pushd( File::Temp::tempdir( CLEANUP => 0 ), $options ) };
    croak $@ if $@;
    $dir->{_tempd} = 1;
    $dir->{_owner} = $$;
    return $dir;
}

#--------------------------------------------------------------------------#
# preserve()
#--------------------------------------------------------------------------#

sub preserve {
    my $self = shift;
    return 1 if !$self->{"_tempd"};
    if ( @_ == 0 ) {
        return $self->{_preserve} = 1;
    }
    else {
        return $self->{_preserve} = $_[0] ? 1 : 0;
    }
}

#--------------------------------------------------------------------------#
# DESTROY()
# Revert to original directory as object is destroyed and cleanup
# if necessary
#--------------------------------------------------------------------------#

sub DESTROY {
    my ($self) = @_;
    my $orig = $self->{_original};
    chdir $orig if $orig; # should always be so, but just in case...
    if ( $self->{_tempd}
        && $self->{_owner} == $$
        && !$self->{_preserve} )
    {
        # don't destroy existing $@ if there is no error.
        my $err = do {
            local $@;
            eval { rmtree( $self->{_pushd} ) };
            $@;
        };
        carp $err if $err;
    }
}

1;

__END__

=head1 SYNOPSIS

 use File::pushd;

 chdir $ENV{HOME};

 # change directory again for a limited scope
 {
     my $dir = pushd( '/tmp' );
     # working directory changed to /tmp
 }
 # working directory has reverted to $ENV{HOME}

 # tempd() is equivalent to pushd( File::Temp::tempdir )
 {
     my $dir = tempd();
 }

 # object stringifies naturally as an absolute path
 {
    my $dir = pushd( '/tmp' );
    my $filename = File::Spec->catfile( $dir, "somefile.txt" );
    # gives /tmp/somefile.txt
 }

=head1 DESCRIPTION

File::pushd does a temporary C<chdir> that is easily and automatically
reverted, similar to C<pushd> in some Unix command shells.  It works by
creating an object that caches the original working directory.  When the object
is destroyed, the destructor calls C<chdir> to revert to the original working
directory.  By storing the object in a lexical variable with a limited scope,
this happens automatically at the end of the scope.

This is very handy when working with temporary directories for tasks like
testing; a function is provided to streamline getting a temporary
directory from L<File::Temp>.

For convenience, the object stringifies as the canonical form of the absolute
pathname of the directory entered.

B<Warning>: if you create multiple C<pushd> objects in the same lexical scope,
their destruction order is not guaranteed and you might not wind up in the
directory you expect.

=head1 USAGE

 use File::pushd;

Using File::pushd automatically imports the C<pushd> and C<tempd> functions.

=head2 pushd

 {
     my $dir = pushd( $target_directory );
 }

Caches the current working directory, calls C<chdir> to change to the target
directory, and returns a File::pushd object.  When the object is
destroyed, the working directory reverts to the original directory.

The provided target directory can be a relative or absolute path. If
called with no arguments, it uses the current directory as its target and
returns to the current directory when the object is destroyed.

If the target directory does not exist or if the directory change fails
for some reason, C<pushd> will die with an error message.

Can be given a hashref as an optional second argument.  The only supported
option is C<untaint_pattern>, which is used to untaint file paths involved.
It defaults to {qr{^(L<-+@\w./>+)$}}, which is reasonably restrictive (e.g.
it does not even allow spaces in the path).  Change this to suit your
circumstances and security needs if running under taint mode. *Note*: you
must include the parentheses in the pattern to capture the untainted
portion of the path.

=head2 tempd

 {
     my $dir = tempd();
 }

This function is like C<pushd> but automatically creates and calls C<chdir> to
a temporary directory created by L<File::Temp>. Unlike normal L<File::Temp>
cleanup which happens at the end of the program, this temporary directory is
removed when the object is destroyed. (But also see C<preserve>.)  A warning
will be issued if the directory cannot be removed.

As with C<pushd>, C<tempd> will die if C<chdir> fails.

It may be given a single options hash that will be passed internally
to C<pushd>.

=head2 preserve

 {
     my $dir = tempd();
     $dir->preserve;      # mark to preserve at end of scope
     $dir->preserve(0);   # mark to delete at end of scope
 }

Controls whether a temporary directory will be cleaned up when the object is
destroyed.  With no arguments, C<preserve> sets the directory to be preserved.
With an argument, the directory will be preserved if the argument is true, or
marked for cleanup if the argument is false.  Only C<tempd> objects may be
marked for cleanup.  (Target directories to C<pushd> are always preserved.)
C<preserve> returns true if the directory will be preserved, and false
otherwise.

=head1 DIAGNOSTICS

C<pushd> and C<tempd> warn with message
C<"Useless use of File::pushd::I<%s> in void context"> if called in
void context and the warnings category C<void> is enabled.

  {
    use warnings 'void';

    pushd();
  }

=head1 SEE ALSO

=for :list
* L<File::chdir>

=cut

# vim: ts=4 sts=4 sw=4 et:


Makefile.PL
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "change directory temporarily for a limited scope",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "File-pushd",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "File::pushd",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Cwd" => 0,
    "Exporter" => 0,
    "File::Path" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "overload" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Basename" => 0,
    "File::Spec" => 0,
    "File::Spec::Functions" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "1.017",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Cwd" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Basename" => 0,
  "File::Path" => 0,
  "File::Spec" => 0,
  "File::Spec::Functions" => 0,
  "File::Temp" => 0,
  "Test::More" => "0.96",
  "overload" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
